File: Project/Student.cpp
#include "Student.hpp"

#include "StringUtils.hpp"

#include <stdexcept>


Student::Student(std::shared_ptr<Group> group, unsigned int studentID, const std::string& firstName, const std::string& lastName, const std::string& middleName) {
    if (group == nullptr) {
        throw std::invalid_argument("Group cannot be nullptr");
    }
    m_group = group;
    m_studentID = studentID;
    SetFirstName(firstName);
    SetLastName(lastName);
    SetMiddleName(middleName);
}

unsigned int Student::GetStudentID() const {
    return m_studentID;
}

void Student::SetStudentID(unsigned int value) {
    m_studentID = value;
}

const std::string& Student::GetFirstName() const {
    return m_firstName;
}

void Student::SetFirstName(const std::string& value) {
    if (StringUtils::ConstainsSpacesOrEmpty(value)) {
        throw std::invalid_argument("First name cannot contain spaces or be empty");
    }
    m_firstName = value;
}

const std::string& Student::GetLastName() const {
    return m_lastName;
}

void Student::SetLastName(const std::string& value) {
    if (StringUtils::ConstainsSpacesOrEmpty(value)) {
        throw std::invalid_argument("Last name cannot contain spaces or be empty");
    }
    m_lastName = value;
}

const std::string& Student::GetMiddleName() const {
    return m_middleName;
}

void Student::SetMiddleName(const std::string& value) {
    if (StringUtils::ConstainsSpacesOrEmpty(value)) {
        throw std::invalid_argument("Middle name cannot contain spaces or be empty");
    }
    m_middleName = value;
}

std::string Student::GetFullName() const {
    return m_firstName + " " + m_lastName + " " + m_middleName;
}

std::string Student::GetLastNameWithInitials() const {
    return m_lastName + " " + m_firstName[0] + ". " + m_middleName[0] + ".";
}

std::weak_ptr<Group> Student::GetGroup() const {
    return m_group;
}

SessionResults& Student::GetSessionResults() {
    return m_sessionResults;
}

File: ProjectCS/Data/Student.cs
using System;

namespace SessionResultsCS
{
    public class Student
    {
        private string _firstName;
        private string _lastName;
        private string _middleName;


        // Номер студенческого билета
        public uint StudentID { get; set; }

        // Имя
        public string FirstName
        {
            get { return _firstName; }
            set
            {
                if (value.Contains(' ') || value.Length == 0)
                {
                    throw new ArgumentException("First name cannot contain spaces or be empty");
                }
                _firstName = value;
            }
        }

        // Фамилия
        public string LastName
        {
            get { return _lastName; }
            set
            {
                if (value.Contains(' ') || value.Length == 0)
                {
                    throw new ArgumentException("Last name cannot contain spaces or be empty");
                }
                _lastName = value;
            }
        }

        // Отчество
        public string MiddleName
        {
            get { return _middleName; }
            set
            {
                if (value.Contains(' ') || value.Length == 0)
                {
                    throw new ArgumentException("Middle name cannot contain spaces or be empty");
                }
                _middleName = value;
            }
        }

        // Группа
        public Group Group { get; }

        // Результаты сессии
        public SessionResults SessionResults { get; }


        public Student(Group group, uint studentID, string firstName, string lastName, string middleName)
        {
            Group = group;
            StudentID = studentID;
            FirstName = firstName;
            LastName = lastName;
            MiddleName = middleName;
            SessionResults = new SessionResults();
        }


        // Возвращает полное имя (фамилию, имя и отчество)
        public string GetFullName()
        {
            return $"{FirstName} {LastName} {MiddleName}";
        }

        // Возвращает фамилию и инициалы
        public string GetLastNameWithInitials()
        {
            return $"{LastName} {FirstName[0]}. {MiddleName[0]}.";
        }
    }
}
